logger/clickhouse.cc

```diff
@@ -107,22 +154,51 @@ TfwClickhouse::handle_block_error() noexcept
 	return false;
 }
 
-TfwClickhouse::TfwClickhouse(const ClickHouseConfig &config)
-	: table_name_(config.table_name),
-	  max_events_(config.max_events)
+bool
+TfwClickhouse::should_attempt_reconnect() const noexcept
 {
-	auto opts = ch::ClientOptions();
+	if (!needs_reconnect.load()) {
+		return false;
+	}
 
-	opts.SetHost(config.host);
-	opts.SetPort(config.port);
-	opts.SetDefaultDatabase(config.db_name);
+	auto now = std::chrono::steady_clock::now();
+	auto last_attempt = last_reconnect_attempt.load();
+	auto timeout = reconnect_timeout.load();
 
-	if (const auto user = config.user.value_or(""); !user.empty())
-		opts.SetUser(user);
-	if (const auto pswd = config.password.value_or(""); !pswd.empty())
-		opts.SetPassword(pswd);
+	return (now - last_attempt) >= timeout;
+}
 
-	client_ = std::make_unique<ch::Client>(opts);
+void
+TfwClickhouse::update_reconnect_timeout(bool success) noexcept
+{
+	if (success) {
+		reconnect_timeout.store(std::chrono::seconds(0));
```

Comment:
We use early exit functions structuring to minimize nesting

```suggestion
		reconnect_timeout.store(std::chrono::seconds(0));
		return;
	}
	....
```